pybop/problems/base_problem.py

```diff
@@ -69,10 +70,11 @@ def __init__(
         self.signal = signal or ["Voltage [V]"]
         self.set_initial_state(initial_state)
         self._dataset = None
-        self._domain_data = None
         self._target = None
         self.verbose = False
         self.failure_output = np.asarray([np.inf])
+        if not hasattr(self, "_domain_data"):
```

Comment:
classes should either have a field or not. If you want you can have an optional field (i.e. can be `None`), but having the field possibly not even exist is prone to unexpected exceptions.